### PR TITLE
ck tile engine integrate with gemm unit tests

### PR DESCRIPTION
**Pull request by msaffari-amd**
## Proposed changes

Based on ckTileEngine, some new unit test has been added to tests that make instances based on config files and test them. 

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

### CODE REVIEW

File: test/ck_tile/gemm/configs/config2.json
Comment (aosewski) [inline test/ck_tile/gemm/configs/config2.json:1]: Why actually you need the config1 and config2 ?

File: test/ck_tile/gemm/gemm_kernel_validation_base.hpp
Comment (aosewski) [inline test/ck_tile/gemm/gemm_kernel_validation_base.hpp:1]: No copyright header.

File: test/ck_tile/gemm/gemm_kernel_validation_base.hpp [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            stride_c,
+        };
+
+        // Compute reference result for validation
+        compute_reference_gemm(a_m_k, b_k_n, c_m_n_host_result);
+
+        try
+        {
+            // Configure stream for kernel execution
+            ck_tile::stream_config stream_cfg{nullptr, true, 0, 1, 1, true, false, 1};
```
Comment (aosewski) [inline test/ck_tile/gemm/gemm_kernel_validation_base.hpp]: No need for warmup & kernel timing. Just single iteration is enough.

```suggestion
            ck_tile::stream_config stream_cfg{nullptr, false, 0, 0, 1};
```

File: test/ck_tile/gemm/test_gemm_instance_validation_bf16.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+};
+
+TYPED_TEST_SUITE(GemmKernelValidationTestBF16, DispatcherTypesBF16);
+
+TYPED_TEST(GemmKernelValidationTestBF16, AllKernelConfigurations)
+{
+    int total_kernels  = 0;
+    int passed_kernels = 0;
+
+    for(bool structured_sparsity : {false, true})
```
Comment (aosewski) [inline test/ck_tile/gemm/test_gemm_instance_validation_bf16.cpp]: How is that this is a run-time parameter ?

File: test/ck_tile/gemm/test_gemm_instance_validation_bf16.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        for(const auto& [trait_name, kernels] : *kernel_map_)
+        {
+            SCOPED_TRACE("Testing trait: " + trait_name +
+                         " structured_sparsity=" + (structured_sparsity ? "true" : "false"));
+
+            for(size_t i = 0; i < kernels.size(); ++i)
+            {
+                SCOPED_TRACE("Kernel index: " + std::to_string(i));
+
+                for(const auto& prob : kTestProblems)
```
Comment (aosewski) [inline test/ck_tile/gemm/test_gemm_instance_validation_bf16.cpp]: This is not a good order of loops. The problem loop should be the outer loops, as you allocate and initialize memory per problem.

File: test/ck_tile/gemm/gemm_kernel_validation_base.hpp [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            permute_vectors_i4x4_b(b_k_n_dev);
+            b_k_n_dev_buf.ToDevice(b_k_n_dev.data());
+        }
+        else
+        {
+            b_k_n_dev_buf.ToDevice(b_k_n.data());
+        }
+
+        // Initialize output buffer to zero
+        c_m_n_dev_buf.SetZero();
```
Comment (aosewski) [inline test/ck_tile/gemm/gemm_kernel_validation_base.hpp]: Can we do this boilerplate code **just once** per problem configuration? If we shuffle memory and initialize it every time, then this would be the most time consuming part of testing.

File: test/ck_tile/gemm/configs/default_config.json
Comment (aosewski) [inline test/ck_tile/gemm/configs/default_config.json:1]: As a goal I hope we would switch **ALL** our tests to be based on ckTileEngine. This way we should test **much much more** configuration parameters. 
Therefore please add many more tile & parameter configurations. From small tile sizes, squarish to tall & skinny etc. Testing all pipelines, all supported matrix instructions etc. 
Those should be extensive testing, but on small problems so it would run fast.

**Review by aosewski — CHANGES_REQUESTED** [review]
Good start, but need some enhancements to achieve comprehensive testing.

File: test/ck_tile/gemm/configs/config2.json
Comment (msaffari-amd) [reply test/ck_tile/gemm/configs/config2.json:1]: The idea of using different configuration files came from our discussions, where we thought it might be beneficial to separate configurations based on value ranges, rather than combining everything into a single JSON file.

File: test/ck_tile/gemm/gemm_kernel_validation_base.hpp
Comment (msaffari-amd) [reply test/ck_tile/gemm/gemm_kernel_validation_base.hpp:1]: Done

File: test/ck_tile/gemm/gemm_kernel_validation_base.hpp (hunk 1 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm/gemm_kernel_validation_base.hpp]: Done

File: test/ck_tile/gemm/test_gemm_instance_validation_bf16.cpp (hunk 2 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm/test_gemm_instance_validation_bf16.cpp]: Yes, it was run time. I changed it to compile time now.

File: test/ck_tile/gemm/test_gemm_instance_validation_bf16.cpp (hunk 3 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm/test_gemm_instance_validation_bf16.cpp]: Thanks. Good point. I changed it. moved it to outer loop and do allocation once pr problem for all kernels.

File: test/ck_tile/gemm/gemm_kernel_validation_base.hpp (hunk 4 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm/gemm_kernel_validation_base.hpp]: I changed it now. there are 2 functions now. create_problem_context for initialize memories and calculate reference once per problem and test_single_kernel that just run kernel for same allocated data and verify them by initial calculated reference.

File: test/ck_tile/gemm/configs/default_config.json
Comment (msaffari-amd) [reply test/ck_tile/gemm/configs/default_config.json:1]: Completely makes sense, and one reason that we designed this unit test framework to make kernel packages for different JSON files was exactly to cover these different cases. If we wanted to put all of them in the same JSON, then the number of kernels would be a very huge number (as we calculated ~150M+ combinations), but now it's controllable.

The current approach allows us to create focused configuration files like:

small_tiles_config.json for edge cases with 16x16, 32x32 tiles
tall_skinny_config.json for aspect ratio testing with 256x32, 512x16 shapes)
etc ...

File: test/ck_tile/gemm/test_gemm_instance_validation_bf16.cpp (hunk 2 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm/test_gemm_instance_validation_bf16.cpp]: Yes, it was run_time. I changed it to compile time now.

File: test/ck_tile/gemm/configs/config1.json [hunk 5]
```diff
@@ -0,0 +1,88 @@
+{
+  "problem": {
+  },
+  "tile_config": {
+    "tile_m": {
+      "values": [
```
Comment (bartekxk) [inline test/ck_tile/gemm/configs/config1.json]: Why we need nested "values" for each cell?

File: test/ck_tile/gemm/test_gemm_instance_validation_fp16.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+// Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
+#include "gemm_kernel_validation_base.hpp"
+#include "gemm_common.hpp"
+#include "gemm_dispatcher.hpp"
+
+using DispatcherTypesFP16 = ::testing::Types<GemmDispatcher>;
+
+// Base template class for FP16 GEMM validation
+template <typename DispatcherType, bool StructuredSparsity>
+class GemmKernelValidationTestFP16Base
```
Comment (bartekxk) [inline test/ck_tile/gemm/test_gemm_instance_validation_fp16.cpp]: Do we need seperate cpp file for each data type?

File: test/ck_tile/gemm/CMakeLists.txt [hunk 7]
```diff
@@ -54,6 +54,129 @@ if(GPU_TARGETS MATCHES "gfx94" OR GPU_TARGETS MATCHES "gfx95" OR GPU_TARGETS MAT
...
+        
+        add_library(test_gemm_instances_${datatype}_${layout}_${config_name} INTERFACE)
+        add_dependencies(test_gemm_instances_${datatype}_${layout}_${config_name} ck_tile_test_gen_${datatype}_${layout}_${config_name})
+        target_include_directories(test_gemm_instances_${datatype}_${layout}_${config_name} INTERFACE
+            ${PROJECT_SOURCE_DIR}/tile_engine/ops/gemm
+            "${working_path}"
+        )
+        set_target_properties(test_gemm_instances_${datatype}_${layout}_${config_name} PROPERTIES LINKER_LANGUAGE CXX)
+
+        add_gtest_executable(test_ck_tile_gemm_instance_validation_${datatype}_${layout}_${config_name}
```
Comment (bartekxk) [inline test/ck_tile/gemm/CMakeLists.txt]: Do we want to keep such cmake complex function instead some friendly C++ api parser?

File: test/ck_tile/gemm/configs/config1.json (hunk 5 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm/configs/config1.json]: This approach is based on how the cktileEngine was originally designed.
The goal was to support running the same algorithms using different JSON configuration files to make new kernels and verify them.
I've now added the capability to use multiple distinct config files, which also helps reduce the level of nesting in test cases, as you pointed out

File: test/ck_tile/gemm/test_gemm_instance_validation_fp16.cpp (hunk 6 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm/test_gemm_instance_validation_fp16.cpp]: Kernel instance libraries are type-specific, meaning that for each data type, kernel code is generated in a separate folder using gemm_instance_builder.py.

We thought this structure might offer a cleaner way to manage the generated kernel instances, allowing each package to be tested independently through its own unit test.

**Review by spolifroni-amd — DISMISSED** [review]

File: test/ck_tile/gemm_tile_engine/configs/simple_test_config.json
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/configs/simple_test_config.json:1]: This is not smoke test in any terms ;) What about `simple_test_config.json`

File: test/ck_tile/gemm_tile_engine/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    
+    # Generate kernel header using tile_engine's Python script
+    add_custom_command(
+        OUTPUT ${test_header}
+        COMMAND ${Python3_EXECUTABLE} ${TILE_ENGINE_GEMM_DIR}/gemm_instance_builder.py
+                --working_path ${working_path}
+                --datatype ${datatype}
+                --layout ${layout}
+                --config_json ${config_json}
+                --gen_single
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/CMakeLists.txt:46]: What is this used for ?

File: test/ck_tile/gemm_tile_engine/test_gemm_simple.cpp [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+#include <gtest/gtest.h>
+#include <iostream>
+
+#include "ck_tile/core.hpp"
+#include "ck_tile/host.hpp"
+#include "tile_engine/ops/gemm/gemm_common.hpp"
+
+// The kernel header is included via compile command line with -include flag
+// It defines SelectedKernel struct, KERNEL_NAME, and tensor data types
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/test_gemm_simple.cpp:15]: Very useful comment! Thanks!

File: test/ck_tile/gemm_tile_engine/test_gemm_simple.cpp [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+                         ::testing::Values(GemmTestParams{256, 256, 128, 1},
+                                           GemmTestParams{256, 256, 1024, 1},
+                                           GemmTestParams{256, 512, 512, 1},
+                                           GemmTestParams{512, 256, 512, 1}),
+                         [](const ::testing::TestParamInfo<GemmTestParams>& param_info) {
+                             return std::to_string(param_info.param.m) + "x" +
+                                    std::to_string(param_info.param.n) + "x" +
+                                    std::to_string(param_info.param.k) + "_splitk" +
+                                    std::to_string(param_info.param.split_k);
+                         });
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/test_gemm_simple.cpp:223]: Nice usage of value-parameterized tests!

**Review by aosewski — COMMENTED** [review]
Great Job!

File: test/ck_tile/gemm_tile_engine/configs/simple_test_config.json
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/simple_test_config.json:1]: You're absolutely right ;). That was just for testing the unit test I developed. I'm now adding meaningful JSON files.

File: test/ck_tile/gemm_tile_engine/test_gemm_simple.cpp (hunk 10 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/test_gemm_simple.cpp:223]: Thanks

File: test/ck_tile/gemm_tile_engine/test_gemm_simple.cpp (hunk 9 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/test_gemm_simple.cpp:15]: Thanks

File: test/ck_tile/gemm_tile_engine/CMakeLists.txt (hunk 8 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/CMakeLists.txt:46]: gemm_instance_builder.py operates in 2 modes:
**--list_kernels**: to find all valid kernels that can be built. it will generates 2 files. 
gemm_kernel_count.txt  that keeps the number of valid kernels that can be built. and
gemm_kernel_list.txt that is the list of these kernels

The other mode is **--gen_single**: It generates actual c++ kernel code.

**Review by aosewski — APPROVED** [review]
